websockify/websocket.py

```diff
@@ -1,1109 +1,825 @@
-            signal.SIGTERM: signal.getsignal(signal.SIGTERM),
-            signal.SIGCHLD: signal.getsignal(signal.SIGCHLD),
-        }
-        signal.signal(signal.SIGINT, self.do_SIGINT)
-        signal.signal(signal.SIGTERM, self.do_SIGTERM)
-        if not multiprocessing:
-            # os.fork() (python 2.4) child reaper
-            signal.signal(signal.SIGCHLD, self.fallback_SIGCHLD)
+
+        b1 = opcode & 0x0f
+        if fin:
+            b1 |= 0x80
+
+        mask_bit = 0
+        if mask_key is not None:
+            mask_bit = 0x80
+            buf = self._mask(buf, mask_key)
+
+        payload_len = len(buf)
+        if payload_len <= 125:
+            header = struct.pack('>BB', b1, payload_len | mask_bit)
+        elif payload_len > 125 and payload_len < 65536:
+            header = struct.pack('>BBH', b1, 126 | mask_bit, payload_len)
+        elif payload_len >= 65536:
+            header = struct.pack('>BBQ', b1, 127 | mask_bit, payload_len)
+
+        if mask_key is not None:
+            return header + mask_key + buf
         else:
-            # make sure that _cleanup is called when children die
-            # by calling active_children on SIGCHLD
-            signal.signal(signal.SIGCHLD, self.multiprocessing_SIGCHLD)
+            return header + buf
 
-        last_active_time = self.launch_time
-        try:
-            while True:
-                try:
-                    try:
-                        startsock = None
-                        pid = err = 0
-                        child_count = 0
-
-                        if multiprocessing:
-                            # Collect zombie child processes
-                            child_count = len(multiprocessing.active_children())
-
-                        time_elapsed = time.time() - self.launch_time
-                        if self.timeout and time_elapsed > self.timeout:
-                            self.msg('listener exit due to --timeout %s'
-                                    % self.timeout)
-                            break
-
-                        if self.idle_timeout:
-                            idle_time = 0
-                            if child_count == 0:
-                                idle_time = time.time() - last_active_time
-                            else:
-                                idle_time = 0
-                                last_active_time = time.time()
-
-                            if idle_time > self.idle_timeout and child_count == 0:
-                                self.msg('listener exit due to --idle-timeout %s'
-                                            % self.idle_timeout)
-                                break
+    def _decode_hybi(self, buf):
+        """ Decode HyBi style WebSocket packets.
+        Returns:
+            {'fin'          : boolean,
+             'opcode'       : number,
+             'masked'       : boolean,
+             'length'       : encoded_length,
```

Comment:
> Huh? It is not the payload length, but rather the length of the entire packet.

whoops, I typoed the comment -- what I meant was it used to be just the payload length, so it should probably be `frame_length` or something like that to not confuse things.

> dicts are very fast, so I doubt we'll save much. :)

Sure, but an named tuple is self-documenting in terms of fields and prevents you from accidentally setting different fields.